Apps/Src/ReadTritium.c

```diff
@@ -127,3 +121,52 @@ float Motor_RPM_Get(){ //getter function for motor RPM
 float Motor_Velocity_Get(){ //getter function for motor velocity
 	return Motor_Velocity;
 }
+
+
+/**
+ * Error handler functions
+ * Passed as callback functions to the main assertTaskError function by assertTritiumError
+*/
+
+/**
+ * @brief A callback function to be run by the main assertTaskError function for hall sensor errors
+ * restart the motor if the number of hall errors is still less than the RESTART_THRESHOLD.
+ */ 
+static void handler_Tritium_HallError(void) {
+	MotorController_Restart(); 
+}
+
+
+/**
+ * @brief   Assert a Tritium error by checking Motor_FaultBitmap
+ * and asserting the error with its handler callback if one exists.
+ *  Can result in restarting the motor (while < RESTART_THRESHOLD number of hall sensor errors)
+ * or locking the scheduler and entering a nonrecoverable fault (all other cases)
+ * @param   motor_err Bitmap with motor error codes to check
+ */
+static void assertTritiumError(tritium_error_code_t motor_err){   
+	static uint8_t hall_fault_cnt = 0; //trip counter, doesn't ever reset
+
+	Error_ReadTritium = motor_err; // Store error code for inspection info
+
+	if(motor_err != T_NONE){ // We need to handle an error
+		if(motor_err != T_HALL_SENSOR_ERR){
+			 // Assert a nonrecoverable error with no callback function- nonrecoverable will kill the motor and infinite loop
+			assertTaskError(OS_READ_TRITIUM_LOC, motor_err, NULL, OPT_LOCK_SCHED, OPT_NONRECOV);
+		
+		}else{
+			hall_fault_cnt++; 
+			
+			//try to restart the motor a few times and then fail out
+			if(hall_fault_cnt > RESTART_THRESHOLD){  
+				// Assert a nonrecoverable error that will kill the motor, display a fault screen, and infinite loop
+				assertTaskError(OS_READ_TRITIUM_LOC, motor_err, NULL, OPT_LOCK_SCHED, OPT_NONRECOV);
+			} else { // Try restarting the motor
+				// Assert a recoverable error that will run the motor restart callback function
+				assertTaskError(OS_READ_TRITIUM_LOC, motor_err, handler_Tritium_HallError, OPT_NO_LOCK_SCHED, OPT_RECOV);
+        	} 
+		}
+	}
+
+	Error_ReadTritium = T_NONE; // Clear the error after handling it
+}
```

Comment:
```suggestion
static void assertTritiumError(tritium_error_code_t motor_err){   
	static uint8_t hall_fault_cnt = 0; //trip counter, doesn't ever reset

	Error_ReadTritium = motor_err; // Store error code for inspection info
	if(motor_err == T_NONE) return; // No error, return
	
	if(motor_err != T_HALL_SENSOR_ERR){
		// Assert a nonrecoverable error with no callback function- nonrecoverable will kill the motor and infinite loop
		assertTaskError(OS_READ_TRITIUM_LOC, motor_err, NULL, OPT_LOCK_SCHED, OPT_NONRECOV);
		return;
	}

	hall_fault_cnt++; 

	//try to restart the motor a few times and then fail out
	if(hall_fault_cnt > RESTART_THRESHOLD){  
		// Assert a nonrecoverable error that will kill the motor, display a fault screen, and infinite loop
		assertTaskError(OS_READ_TRITIUM_LOC, motor_err, NULL, OPT_LOCK_SCHED, OPT_NONRECOV);
		return;
	} 
	// Try restarting the motor
	// Assert a recoverable error that will run the motor restart callback function
	assertTaskError(OS_READ_TRITIUM_LOC, motor_err, handler_Tritium_HallError, OPT_NO_LOCK_SCHED, OPT_RECOV); 
	
	Error_ReadTritium = T_NONE; // Clear the error after handling it
}
```

Deeply nested if-elses like this make the code harder to comprehend, and limits the maintenance of the code. Often-times when you have these situations, it means your "happy path" (shortest return path) isn't the one you built the function with in mind. Generally ,to fix this, you can just flip outer if conditions, and turn them into early returns. This puts your "happy path" at the top, getting out of the way early, and you can focus on what the function is meant to do in specific cases closer to the bottom. This makes the code much simpler to read, as it is very obvious which combination/set of conditions leads to an early return, and the default case is to do an action.

Reply:
I'd like to make the function clearer, but I wrote it by the priority of each action, so I don't know how I would move things around. For instance, locking the scheduler should come first, and an infinite loop should go last. It's also sort of a mix-and-match situation where we could choose not to lock the scheduler but still have a callback, so I don't know if there's really a "happy path," sadly.

Reply:
Don't worry about action priority. Readability (in this case) is king. Happy path == shortest sequence of steps / return path. Deeply nested if-else is heavily frowned upon in industry.

Reply:
Why do you respond so fast? 😭 I'm starting to think you really do have a drug-style addiction to LHR

Reply:
Life was simpler on controls, I try to visit when I can 😉

Reply:
@KnockbackNemo @IshDeshpa my suggested code block here should do the correct action, any reason it hasn't been accepted in?

Reply:
Nope that's my bad, it looks great